### PR TITLE
minor refactor and optimize serviceregistry

### PR DESCRIPTION
**Pull request by wulianglongrd**
Some miscellaneous refactorings and optimizations:
1. ~fix typo: `initialSyncTimedout` -> `initialSyncTimeout`~
2. remove redundant set options, remove `ConvertServiceEntry` function
3. extract a `isDNSTypeServiceEntry` function
4. set capacity when initializing slices in some places
5. for other information, see code comments

### CODE REVIEW

File: pilot/pkg/serviceregistry/serviceentry/conversion.go [hunk 1]
```diff
@@ -320,23 +322,29 @@ func (s *Controller) convertWorkloadEntryToServiceInstances(wle *networking.Work
...
 	if serviceEntry == nil {
 		return nil
 	}
 	if services == nil {
 		services = convertServices(cfg)
 	}
+
+	endpointsNum := len(serviceEntry.Endpoints)
+	hostnameToServiceInstance := false
+	if len(serviceEntry.Endpoints) == 0 && serviceEntry.WorkloadSelector == nil && isDNSTypeServiceEntry(serviceEntry) {
```
Comment (wulianglongrd) [inline pilot/pkg/serviceregistry/serviceentry/conversion.go:335]: It's moved up, we don't need to check every time.

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 2]
```diff
@@ -210,9 +210,9 @@ type Controller struct {
 
 	ambientIndex
 
-	// initialSyncTimedout is set to true after performing an initial processing timed out.
-	initialSyncTimedout *atomic.Bool
-	meshWatcher         mesh.Watcher
+	// initialSyncTimeout is set to true after performing an initial processing timed out.
+	initialSyncTimeout *atomic.Bool
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:214]: revert, this is to indicate it is a bool

File: pilot/pkg/serviceregistry/serviceentry/controller.go [hunk 3]
```diff
@@ -777,15 +759,15 @@ func (s *Controller) doEdsUpdate(keys sets.Set[instancesKey]) {
 
 // buildEndpoints builds endpoints for the instance keys.
 func (s *Controller) buildEndpoints(keys map[instancesKey]struct{}) map[instancesKey][]*model.IstioEndpoint {
-	var endpoints map[instancesKey][]*model.IstioEndpoint
-	allInstances := []*model.ServiceInstance{}
+	allInstances := make([]*model.ServiceInstance, 0, len(keys))
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/serviceentry/controller.go:762]: This is not correct, below i can be not only one instance

File: pilot/pkg/serviceregistry/serviceentry/conversion.go [hunk 4]
```diff
@@ -179,7 +180,8 @@ func convertServices(cfg config.Config) []*model.Service {
 	if serviceEntry.WorkloadSelector != nil {
 		labelSelectors = serviceEntry.WorkloadSelector.Labels
 	}
-	hostAddresses := []*HostAddress{}
+
+	hostAddresses := make([]*HostAddress, 0, len(serviceEntry.Hosts)*slices.NonZeroLenOr(serviceEntry.Addresses, 1))
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/serviceentry/conversion.go:184]: we donot need such complexity, unless this is the perf bottle neck

File: pkg/maps/maps.go [hunk 5]
```diff
@@ -75,3 +77,13 @@ func Contains[M1, M2 ~map[K]V, K comparable, V comparable](superset M1, subset M
 func EqualFunc[M1 ~map[K]V1, M2 ~map[K]V2, K comparable, V1, V2 any](m1 M1, m2 M2, eq func(V1, V2) bool) bool {
 	return maps.EqualFunc(m1, m2, eq)
 }
+
+// CountSliceValue count the number of elements in the map value.
+// This value is []any type.
+func CountSliceValue[M ~map[K][]V, K comparable, V any](m M) int {
```
Comment (hzxuzhonghu) [inline pkg/maps/maps.go:83]: This is not a fundamental helper,

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 2 above)
Comment (wulianglongrd) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:214]: I didn't realize it had this meaning at all, how did you know?

File: pilot/pkg/serviceregistry/serviceentry/controller.go (hunk 3 above)
Comment (wulianglongrd) [reply pilot/pkg/serviceregistry/serviceentry/controller.go]: reverted

File: pilot/pkg/serviceregistry/serviceentry/conversion.go (hunk 4 above)
Comment (wulianglongrd) [reply pilot/pkg/serviceregistry/serviceentry/conversion.go]: reverted

File: pkg/maps/maps.go (hunk 5 above)
Comment (wulianglongrd) [reply pkg/maps/maps.go]: moved out

**Comment by hzxuzhonghu** [comment]
why the ci tests does not run?

**Comment by dhawton** [comment]
/retest

> why the ci tests does not run?

Prow issues https://istio.slack.com/archives/C6FCV6WN4/p1716830124853679

**Review by howardjohn — APPROVED** [review]

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**